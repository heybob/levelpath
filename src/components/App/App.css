.App {
  text-align: center;
  margin: 0;
  min-width: 785px;
}

.list-container {
  display: flex;
  flex-direction: column;
  justify-items: center;
  align-items: center;
}

.header-container {
  position: sticky;
  top: 0;
  margin: 0;
  background-color: white;
  border-bottom: 1px solid black;
  margin-bottom: 2rem;
}