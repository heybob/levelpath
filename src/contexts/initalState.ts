import { Text } from "../shared/constants"

export const initialState = {
    isLoading: false, 
    data: [[]], 
    paginatedData: [], 
    pageNum: 1, 
    totalPages: 0,
    pageSize: 50, 
    today: new Date('December 17, 1995 03:24:00'),
    apiError: false,
    pageTitle: Text.headerTitle
}