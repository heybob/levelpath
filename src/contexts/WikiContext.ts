import React from "react";
import { createContext } from "react";
import { initialState } from "./initalState";
import { WikiState } from "../shared/types";

export const  WikiContext = createContext<{ 
    state: WikiState; dispatch: React.Dispatch<any>;}>(
        {state: initialState, dispatch: () => null});