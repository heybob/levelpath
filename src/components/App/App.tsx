import { useEffect, useContext, useCallback } from 'react';
import './App.css';
import { loadWikiInProgress, loadWikiFailure, loadWikiSuccess, pageChange, setTodaysDate } from './actions';
import { getWikiData } from '../../shared/api';
import { WikiContext } from '../../contexts/WikiContext';
import {WikiResponse } from '../../shared/types';
import { processWikiBirthsData } from '../../shared/utilities';
import { Header } from '../Header/Header';
import { Paginator } from '../Paginator/Paginator';
import { LoadingIndicator } from '../LoadingIndicator/LoadingIndicator';
import { ErrorModal } from '../ErrorModal/ErrorModal';
import { ProfileList } from '../ProfileList/ProfileList';
import {Text} from '../../shared/constants';
import { Button } from '../Button/Button';


export function App() {
  const {state, dispatch}  = useContext(WikiContext);
  const {data, paginatedData, isLoading, apiError, pageNum, totalPages, today, pageSize} = state;

  useEffect(()=> {
    dispatch(setTodaysDate(new Date()));
  },[dispatch]);
  
  const getWikiBirthsData = async () => {
    try {
      dispatch(loadWikiInProgress());
      const month = String(today.getMonth() + 1).padStart(2,'0');
      const day = String(today.getDate()).padStart(2,'0');
      const response = await getWikiData(month, day);
      const wikiData: WikiResponse = await response.json();
      const wikiDataTotalPages = Math.ceil(wikiData.births.length / pageSize);
      const profileData = processWikiBirthsData(wikiData, wikiDataTotalPages, pageSize);
      dispatch(loadWikiSuccess({data: profileData, pageNum: pageNum, totalPages: wikiDataTotalPages}));
    } catch(e){
        dispatch(loadWikiFailure(e));
    }
  }

  async function handleClick() {
    getWikiBirthsData();
  }
  
  const handlePageChange = useCallback((pNum: number) => {
    if(pNum !== pageNum){
      dispatch(pageChange(pNum));
    }
  }, [pageNum, dispatch]);

  return (
    <div className="App">
      <div className='header-container'>
      <Header />
      {totalPages ? <Paginator handlePageChange={handlePageChange} /> : null}
      </div>
      {!data[0].length ? <Button name={Text.getDataButton} clickHandler={handleClick} disabled={isLoading} /> : null}
      {isLoading ? <LoadingIndicator /> :  null}
      {apiError ? <ErrorModal /> : null}
      {paginatedData.length ? <ProfileList /> : null}
    </div>
  );
}

export default App;
