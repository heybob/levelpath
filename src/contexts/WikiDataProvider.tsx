import React from 'react'
import { useReducer} from "react";
import { wikiReducer } from "../components/App/reducer";
import { initialState } from "./initalState";
import { WikiContext } from "./WikiContext";

type Props = {
    children: React.ReactNode
  }

export function WikiDataProvider({children}: Props) {

    const [state, dispatch] = useReducer(wikiReducer, initialState);

    return (
        <WikiContext.Provider value={{state, dispatch}}>{children}</WikiContext.Provider>
    );

}