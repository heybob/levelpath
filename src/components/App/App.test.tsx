import renderer from 'react-test-renderer';
import { WikiDataProvider } from '@App/contexts/WikiDataProvider';
import { App } from '@App/components/App/App';

describe('Testing App Component', () => {
  it('It should render App', () => {
      const tree = renderer.create(
      <WikiDataProvider>
            <App />
        </WikiDataProvider>
      ).toJSON()
    expect(tree).toMatchSnapshot();
  });
});