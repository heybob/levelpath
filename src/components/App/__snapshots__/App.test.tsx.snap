// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing App Component It should render App 1`] = `
<div
  className="App"
>
  <div
    className="header-container"
  >
    <div
      className="header"
    >
      <h1>
        Wikipedia Presents: Born On This Day
         
      </h1>
      <h2>
        12
        /
        17
        /
        1995
      </h2>
    </div>
  </div>
  <button
    disabled={false}
    onClick={[Function]}
  >
    Click to See Today's Birthdays!
  </button>
</div>
`;
